### PR TITLE
Improve `vec_equal()` performance

### PR DESCRIPTION
**Pull request by lionel-**
This pulls the dereference of atomic SEXP arrays out of the `vec_equal()` loop to improve performance. A new `vctrs_type` enum supports the genericity over types.

I have tried several ways:

- `no_deref`: Pass a reference object to `equal_scalar()` that contains an array pointer if applicable:

    ```c
    struct vctrs_ref {
      SEXP object;
      enum vctrs_type type;
      void* pointer;
      R_len_t i;
    };
    ```

  This approach improved performance but not enough.

- `no_branch`: Same as above, but pulls the `if (na_equal)` branch out of the loop. This had no impact.

- `template`: switch over vctrs types with the same loop template. This is the approach that is implemented in this PR.

- `template_full`: Like `template` but with no `na_equal` branch. Again, no impact.

faster-loop-desktop

The `dotcall` case directly `.Call()`s into `vctrs_equal()` for better comparison with base R's `==`.

### CODE REVIEW

File: src/equal.c [hunk 1]
```diff
@@ -1,68 +1,72 @@
 #include "vctrs.h"
 
-int equal_scalar(SEXP x, int i, SEXP y, int j, bool na_equal) {
-  if (TYPEOF(x) != TYPEOF(y))
-    return false;
+int lgl_equal_scalar(int* x, int* y, bool na_equal) {
+  if (*x == NA_LOGICAL) return na_equal ? *y == NA_LOGICAL : NA_LOGICAL;
```
Comment (hadley) [inline src/equal.c]: I wonder if it's worth rewriting to something like:

```R
if (na_equal) {
  return *x == *y 
} else {
  return (*x == NA_LOGICAL || *y == NA_LOGICAL) ? NA_LOGICAL : *x == *y
}
```

I think that's easier to understand, and you're in a good position to see if it has any performance implications.

Comment (hadley) [reply src/equal.c]: I have a vague feeling it might be worth re-checking if there's any difference in performance pulling the `na_equal` comparison after this change..... Now I've read further down, that would clearly introduce a bunch of duplication, so it's probably not worth trying for now.

File: src/types.c [hunk 2]
```diff
@@ -11,3 +11,22 @@ bool is_record(SEXP x) {
 bool is_scalar(SEXP x) {
   return Rf_inherits(x, "vctrs_sclr");
 }
+
+enum vctrs_type vec_typeof(SEXP x) {
```
Comment (hadley) [inline src/types.c]: Looks good, and this will be straight forward to extend to other base vector types (e.g. factor, Date, ...)

File: src/vctrs.h [hunk 3]
```diff
@@ -17,6 +35,8 @@ bool is_scalar(SEXP x);
 int equal_object(SEXP x, SEXP y, bool na_equal);
 bool equal_names(SEXP x, SEXP y);
 
+// Caller must ensure proper types and sizes. These functions are meant
```
Comment (hadley) [inline src/vctrs.h]: It's not clear to me what "proper types and sizes" are. Can you be more specific?

File: src/equal.c (hunk 1 above)
Comment (lionel-) [reply src/equal.c]: I tried this in the `template_full` case (now `template_no_branch` in the new graphic) and it made no difference. I expected to see some difference as well, especially with a high proportion of missing values.

File: src/vctrs.h (hunk 3 above)
Comment (lionel-) [reply src/vctrs.h]: This is now:

```c
// Caller must ensure proper SEXPTYPEs and lengths for `x` and `y`.
// These functions are meant to be used in loops. These must check
// bounds.
```

**Comment by lionel-** [comment]
I realised I didn't check performance when the vector contains missing values 🤦.

Interestingly base R was faster with a propertion of missing values close to 50% in the `NA` propagation case. I fixed this gap by recording the element values on the stack: https://github.com/r-lib/vctrs/pull/224/commits/ec10c7be9754ec3e87f660e7a650a9a734c0765a. I'm not sure why the compiler doesn't do this as an optimisation. Assembler output confirms that the pointer are derefed repeatedly when the values are not explicitly saved on the stack. Perhaps because of aliasing? But there is no operation that could change the pointed values...

Here is a more complete comparison (all of these directly `.Call()` into their implementation):

- `template2` uses your suggested control flow
- `template2_no_alias` is the same but with values saved on the stack

faster-loop-desktop

File: src/equal.c [hunk 4]
```diff
@@ -1,68 +1,92 @@
...
-  if (TYPEOF(x) != TYPEOF(y))
-    return false;
-
-  switch(TYPEOF(x)) {
-  case LGLSXP: {
-    int xi = LOGICAL(x)[i], yj = LOGICAL(y)[j];
-    if (xi == NA_LOGICAL) return na_equal ? yj == NA_LOGICAL : NA_LOGICAL;
-    if (yj == NA_LOGICAL) return na_equal ? xi == NA_LOGICAL : NA_LOGICAL;
+int lgl_equal_scalar(int* x, int* y, bool na_equal) {
+  int xi = *x;
```
Comment (hadley) [inline src/equal.c]: Can you add a comment here noting that this is a performance optimisation?

File: src/vctrs.h [hunk 5]
```diff
@@ -17,6 +35,9 @@ bool is_scalar(SEXP x);
 int equal_object(SEXP x, SEXP y, bool na_equal);
 bool equal_names(SEXP x, SEXP y);
 
+// Caller must ensure proper SEXPTYPEs and lengths for `x` and `y`.
```
Comment (hadley) [inline src/vctrs.h]: I still don't quite get what you mean by this. Do you mean that `x` and `y` must have the same `SEXPTYPE`? And that `i` and `j` must be between 0 and `length(x) - 1`/`length(y) - 1`?

Comment (lionel-) [reply src/vctrs.h]: Exactly. How about:

```c
// Callers must ensure identical SEXPTYPEs for `x` and `y` and lengths
// greater than `i` and `j` respectively. These functions are meant to
// be used in loops which must check bounds.
```

Comment (hadley) [reply src/vctrs.h]: I like that! I'd reword slightly:

These functions are meant to be used in loops so it is the callers responsibility to ensure that:
* `x` and `y` have identical SEXTYPES
* `i` is a valid index into `x`, and `j` is a valid index into `y`.
The behaviour is undefined if these conditions are not true.

Comment (lionel-) [reply src/vctrs.h]: Thanks, done!
